ibm/resource_ibm_cis_waf_rule.go

```diff
@@ -0,0 +1,208 @@
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "CIS Domain ID",
+				DiffSuppressFunc: suppressDomainIDDiff,
+			},
+			cisWAFRuleID: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "CIS WAF Rule id",
+			},
+			cisWAFRulePackageID: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "CIS WAF Rule package id",
+			},
+			cisWAFRuleMode: {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "CIS WAF Rule mode",
+				ValidateFunc: InvokeValidator(ibmCISWAFRule, cisWAFRuleMode),
+			},
+			cisWAFRuleDesc: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "CIS WAF Rule descriptions",
+			},
+			cisWAFRulePriority: {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "CIS WAF Rule Priority",
+			},
+			cisWAFRuleGroup: {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "CIS WAF Rule group",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						cisWAFRuleGroupID: {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "waf rule group id",
+						},
+						cisWAFRuleGroupName: {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "waf rule group name",
+						},
+					},
+				},
+			},
+			cisWAFRuleAllowedModes: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "CIS WAF Rule allowed modes",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func resourceIBMCISWAFRuleValidator() *ResourceValidator {
+
+	validateSchema := make([]ValidateSchema, 1)
+	modes := "on, off, default, disable, simulate, block, challenge"
+
+	validateSchema = append(validateSchema,
+		ValidateSchema{
+			Identifier:                 cisWAFRuleMode,
+			ValidateFunctionIdentifier: ValidateAllowedStringValue,
+			Type:                       TypeString,
+			Required:                   true,
+			AllowedValues:              modes})
+	ibmCISWAFRuleValidator := ResourceValidator{ResourceName: ibmCISWAFRule, Schema: validateSchema}
+	return &ibmCISWAFRuleValidator
+}
+
+func resourceIBMCISWAFRuleUpdate(d *schema.ResourceData, meta interface{}) error {
+	cisClient, err := meta.(ClientSession).CisWAFRuleClientSession()
+	if err != nil {
+		return err
+	}
+
+	crn := d.Get(cisID).(string)
+	zoneID, _, err := convertTftoCisTwoVar(d.Get(cisDomainID).(string))
+	cisClient.Crn = core.StringPtr(crn)
+	cisClient.ZoneID = core.StringPtr(zoneID)
+	ruleID := d.Get(cisWAFRuleID).(string)
+	packageID, _, _, _ := convertTfToCisThreeVar(d.Get(cisWAFRulePackageID).(string))
+
+	if d.HasChange(cisWAFRuleMode) {
+		mode := d.Get(cisWAFRuleMode).(string)
+
+		getOpt := cisClient.NewGetWafRuleOptions(packageID, ruleID)
+		getResult, getResponse, err := cisClient.GetWafRule(getOpt)
```

Comment:
Mode is provided by user.. why is get performed here?

Reply:
Need to find owasp or cis mode.  Based on that, we need to set the mode. That's why I read the waf rule to find the mode.